src/planet-a/SubdivTriangleMesh.h

```diff
@@ -0,0 +1,487 @@
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "SubdivSkeleton.h"
+
+#include <osp/types.h>
+
+#include <memory>
+#include <variant>
+
+namespace planeta
+{
+
+// ID for all chunks, from 0 to m_chunkMax
+enum class ChunkId : uint16_t {};
+
+// ID for all shared vertices; from 0 to m_sharedMax
+enum class SharedVrtxId : uint32_t {};
+
+// Indices of vertices, unaware of vertex size;
+// from 0 to (m_chunkMax*m_chunkVrtxCount + m_sharedMax)
+enum class VertexId : uint32_t {};
+
+// IDs for any chunk's shared vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalSharedId : uint16_t {};
+
+// IDs for any chunk's fill vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalFillId : uint16_t {};
+
+using SharedVrtxStorage_t = osp::IdRefCount<SharedVrtxId>::Storage_t;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Convert XY coordinates to a triangular number index
+ *
+ *  0
+ *  1  2
+ *  3  4  5
+ *  6  7  8  9
+ * 10 11 12 13 14
+ * ...
+ * x = right; y = down; (0, 0) = 0
+ *
+ * @param x [in]
+ * @param y [in]
+ * @return triangular number
+ */
+constexpr int xy_to_triangular(int x, int y) noexcept
+{
+    return y * (y + 1) / 2 + x;
+};
+
+constexpr std::pair<ChunkLocalSharedId, bool> coord_to_shared(
+        uint16_t x, uint16_t y, uint16_t chunkEdgeVrtxCount) noexcept
+{
+    // Tests if (x,y) lies along right, bottom, or left edges of triangle
+    // branchless because why not
+    bool lft = (x == 0);
+    bool btm = !lft && (y == chunkEdgeVrtxCount);
+    bool rte = !btm && (x == y);
+
+    // Calculate chunk local shared ID. starts at 0 from the top of the chunk,
+    // and increases counterclockwise along the edge
+    uint32_t offset = lft * y
+                    + btm * (chunkEdgeVrtxCount + x)
+                    + rte * (chunkEdgeVrtxCount * 3 - x);
+
+    return {ChunkLocalSharedId(offset), rte || btm || lft};
+}
+
+//-----------------------------------------------------------------------------
+
+
+struct Chunk
+{
+    SkTriStorage_t m_skeletonTri;
+};
+
+class ChunkedTriangleMeshInfo
+{
+
+public:
+
+    // from wolfram alpha: "sequence 1 3 9 21 45 93"
+    // non-trivial to calculate. ~10 is over the practical limit for
+    // subdivision levels.
+    static inline constexpr std::array<uint16_t, 10> smc_minFansVsLevel =
+    {
+        1, 3, 9, 21, 45, 93, 189, 381, 765, 1533
+    };
+
+    /**
+     * @brief ChunkedTriangleMesh
+     *
+     * @param chunkMax     [in] Max number of chunks
+     * @param subdivLevels [in] Number of times a chunk triangle is subdivided.
+     *                          Practical limit: ~8
+     * @param sharedMax    [in] Max number of shared vertices
+     * @param fanMax       [in] Max number of Fan triangles reserved for a chunk
+     */
+    ChunkedTriangleMeshInfo(uint16_t chunkMax, uint8_t subdivLevels,
+                            uint32_t sharedMax, uint16_t fanMax)
```

Comment:
Recommend giving a brief justification for why these specific int types were chosen for each variable.

I see that you're claiming that more than 8 subdivisions doesn't make sense, so using 8bits is reasonable.

Reply:
A brief justification kind of requires the reader to know about the internals, or what the planet mesh looks like when rendered. In a way, the types kind of hint at how large the numbers should be.